jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -18,721 +18,535 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+        if (
+          target.status === 0 &&
+          target.response === '' &&
+          target.responseText === ''
+        ) {
+          // Server closed the connection, retry and increment the counter.
+          if (retryCounter < 5) {
+            retryCounter += 1;
+            generateRequest(retryCallback);
+            return;
+          }
+          // We have retried 5 times so fall through to call handler.
         }
-        var nonce = nacl.util.decodeBase64(caveatIdObj.Nonce);
-        var firstPartyPub = nacl.util.decodeBase64(
-          caveatIdObj.FirstPartyPublicKey
-        );
-        if(nonce.length !== this.nonceLen) {
-          throw new Exception('bad nonce length');
+        const error = this._getError(target);
+        if (error) {
+          // The interaction failed.
+          onFailure('cannot interact: ' + this._getErrorMessage(error));
+          return;
         }
-        var sealed = nacl.util.decodeBase64(caveatIdObj.Id);
-        var unsealed = nacl.box.open(sealed, nonce, firstPartyPub,
-                                     myKeyPair.secretKey);
-
-        var unsealedStr = nacl.util.encodeUTF8(unsealed);
-        var plain = JSON.parse(unsealedStr);
-        if(plain.Condition === undefined) {
-          throw new Exception('empty condition in third party caveat');
+        // The interaction succeeded.
+        onSuccess(response);
+      };
+      generateRequest(retryCallback);
+    }
+
+    /**
+      Return any error present in the given response.
+
+      @param {Object} target The XHR response.target.
+      @return {Object or String} The error as found in the request.
+    */
+    _getError(target) {
+      // Check bakery statuses.
+      if (
+        target.status !== STATUS_UNAUTHORIZED &&
+        target.status !== STATUS_PROXY_AUTH_REQUIRED
+      ) {
+        return null;
+      }
+      // Bakery protocol errors always have JSON payloads.
+      if (target.getResponseHeader('Content-Type') !== JSON_CONTENT_TYPE) {
+        return null;
+      }
+      // At this point it should be possible to decode the error response.
+      let error;
+      try {
+        error = JSON.parse(target.responseText);
+      } catch(err) {
+        return 'cannot parse error response';
+      }
+      return error;
+    }
+
+    /**
+      Try to parse the given JSON decoded response in order to retrieve a
+      human friendly error.
+
+      @param {Object} jsonResponse The JSON decoded response text.
+      @return {String} The error message.
+    */
+    _getErrorMessage(jsonResponse) {
+      return (
+        jsonResponse.Message ||
+        jsonResponse.message ||
+        jsonResponse.Error ||
+        jsonResponse.error ||
+        'unexpected error: ' + JSON.stringify(jsonResponse)
+      );
+    }
+
+  };
+
+  /**
+    A storage for the macaroon bakery.
+
+    The storage is used to persist macaroons.
+  */
+  const BakeryStorage = class BakeryStorage {
```

Comment:
I'd prefer if this was in its own file, even if it is small and required by the bakery...single class/file old habit and all that ;)